flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/MultipleInputChainingOutput.java

```diff
@@ -0,0 +1,138 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.runtime.tasks;
+
+import org.apache.flink.metrics.Counter;
+import org.apache.flink.metrics.Gauge;
+import org.apache.flink.metrics.SimpleCounter;
+import org.apache.flink.runtime.metrics.groups.OperatorIOMetricGroup;
+import org.apache.flink.runtime.metrics.groups.OperatorMetricGroup;
+import org.apache.flink.streaming.api.operators.Input;
+import org.apache.flink.streaming.api.watermark.Watermark;
+import org.apache.flink.streaming.runtime.metrics.WatermarkGauge;
+import org.apache.flink.streaming.runtime.streamrecord.LatencyMarker;
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
+import org.apache.flink.streaming.runtime.streamstatus.StreamStatusProvider;
+import org.apache.flink.util.OutputTag;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+
+class MultipleInputChainingOutput<T> implements WatermarkGaugeExposingOutput<StreamRecord<T>> {
```

Comment:
In an easy way, it jus duplicates two trivial `collect` methods (3 lines of code each). For the other methods, it's only partially duplicating them, and avoiding code duplication there would introduce even more abstract methods and more mangled code.

I think complicating the hierarchy to avoid this duplication would do more harm than good. Especially that the `ChainingOutput` hasn't changed since 2017.

Reply:
I ran diff on them and see the difference in only `setKeyContextElement1` vs `setKeyContextElement` and in `close`. 

I think the root cause is that `Input` duplicates `OneInputStreamOperator`. And the proper solution would be to refactor `SourceOperator`. But that's apparently out of scope.

Instead, we could use an adapter from operator to input and use only `MultipleInputChainingOutput`, like this:
```
// instead of new ChainingOutput()
currentOperatorOutput = new MultipleInputChainingOutput<>(Input.from(operator), ((OperatorMetricGroup) operator.getMetricGroup()), this, outputTag) {};
// where Input.from returns some Input delegating to operator
```
(I think the delegation overhead will be eliminated by JVM, but it can be tested).

Just a suggestion.

Reply:
> (I think the delegation overhead will be eliminated by JVM, but it can be tested).

Would it? JVM is terrible with callsite optimisations, so I think this would add another virtual call.

But let me try out unifying `Input` with `OneInputStreamOperator`.

Reply:
Done. Please take a look at the commits:

- [FLINK-18905][task/datastream] Convert OneInputStreamOperator to Input
- fixup! [FLINK-18905][task] Allow SourceOperator chaining with MultipleInputStreamTask

Reply:
Great, thanks!